### ISSUE TITLE
Concurrency and Mac M1

### ISSUE THREAD

**Issue by nagman** [lovell/sharp#3182]
## Question about an existing feature

There's this page about concurrency: https://sharp.pixelplumbing.com/api-utility#concurrency.  
But no matter what value I pass to it, this doesn't change anything.

So I wondered if either I didn't understand the meaning of this feature, either it just doesn't work on my device (a Mac M1).

### What are you trying to achieve?

I have a bunch of 18 times 300 images (big 2560x1440 PNG in full quality), and I have to convert them into 3 sizes and 3 formats.  
So an input of 5400 images and an output of 48 600 images.

One input image takes 1.8 s to be processed to the 9 outputs.  
It would take 24 hours to process the whole stuff.

I need a way to speed up my script. So I first thought of multi-threading.  
My first attempt was to leverage node workers, by manually spawning 7 processes in parallel (cause I have 8 cores).  
**But the time it took to process 20 images was exactly the same than when I was doing it on single-thread.**  
Still very odd to me.

Then I found this method: https://sharp.pixelplumbing.com/api-utility#concurrency  
But no matter the value I gave it (and the value was updated, I checked it), it still didn't change anything.

So in recap:
- single-thread with concurrency 1 => same time
- multi-thread with concurrency 1 => same time
- single-thread with concurrency 8 => same time
- multi-thread with concurrency 8 => same time

Actually, it was a little bit slower on multi-thread (maybe the time to spawn a new process).

### When you searched for similar issues, what did you find that might be related?

Just links to this page: https://sharp.pixelplumbing.com/api-utility#concurrency

And this unresolved thread: https://github.com/lovell/sharp/issues/2538

But no explanation about how sharp behaves under the hood.

Is sharp natively multi-threaded?  
Is libvips, on which sharp is built, natively multi-threaded?  
Is multi-threading supported on Mac M1?

And if multi-threading is already handled by `concurrency`, I assume that spawning node workers won't change a thing about performance, am I right?

### Please provide a minimal, standalone code sample, without other dependencies, that demonstrates this question

Here's my repo: https://github.com/nagman/compress-images

### Please provide sample image(s) that help explain this question

Here are my tests:
```
concurrency: 1
nbr of workers: 8

_2.png | worker 5 | 7.4 s
_3.png | worker 1 | 7.5 s
_5.png | worker 3 | 7.5 s
_1.png | worker 2 | 7.7 s
_6.png | worker 7 | 9.7 s
_4.png | worker 6 | 11.0 s

Finished in 12 s
```
```
concurrency: 8
nbr of workers: 8

_1.png | worker 4 | 2.5 s
_5.png | worker 3 | 6.2 s
_2.png | worker 1 | 7.0 s
_6.png | worker 6 | 7.8 s
_3.png | worker 2 | 8.8 s
_4.png | worker 5 | 10.4 s

Finished in 11 s
```
```
concurrency: 8
nbr of workers: 1

_1.png | worker 1 | 2.0 s
_2.png | worker 1 | 1.9 s
_3.png | worker 1 | 1.9 s
_4.png | worker 1 | 1.8 s
_5.png | worker 1 | 1.9 s
_6.png | worker 1 | 1.8 s

Finished in 11 s
```
```
concurrency: 1
nbr of workers: 1

_1.png | worker 1 | 2.0 s
_2.png | worker 1 | 2.0 s
_3.png | worker 1 | 1.8 s
_4.png | worker 1 | 1.8 s
_5.png | worker 1 | 1.8 s
_6.png | worker 1 | 1.8 s

Finished in 11 s
```

**Comment by lovell** [comment]
As you've already seen in the documentation for https://sharp.pixelplumbing.com/api-utility#concurrency

> The maximum number of images that can be processed in parallel is limited by libuv's `UV_THREADPOOL_SIZE` environment variable.


> `concurrency` controls "the number of threads libvips' should create to process each image."

An M1 has 8 cores, which means you'll probably want to process a maximum of 8 images in parallel, so setting `UV_THREADPOOL_SIZE=8` will help achieve this.

https://nodejs.org/api/cli.html#uv_threadpool_sizesize

The default value of `concurrency` will already be 8, although this setting will help much less as the bottleneck for processing PNG images will probably be encoding time, which is single-threaded anyway.

Speeding up PNG compression can be achieved by reducing `effort` and ensuring `palette` and its associated options are *not* enabled.

https://sharp.pixelplumbing.com/api-output#png

**Comment by nagman** [comment]
Oohh! That's a little clearer. Thanks for the answer.

I thought `UV_THREADPOOL_SIZE` was the same thing as `concurrency`, but it seems it's not. The first one defines how many images are processed in parallel, and the second one defines how much CPU cores are dedicated to one image, am I right?

But I have a rough time understanding what's happening under the hood with `UV_THREADPOOL_SIZE`.  
`concurrency` seems pretty straightforward: if I set it to 8, I'll have 8 threads used by sharp.  
But does `UV_THREADPOOL_SIZE` has anything to see with threads or cores?

In the meantime, I've been able to do some more tests and I understood that sharp sets up a queue when you call it.

So I simplified a lot my script and I removed all the node workers and chunks setup in order to just have one big array of 48 600 promises (each promise being a sharp operation).

Instead of running 48 600 operations in parallel, sharp puts them in one big queue and treats them by small chunks.  So my computer never gets overwhelmed. Each operation takes 0.16 s in average, which is pretty fast, and as I have 9 operations per image, it sums up to about 1.6 s per image (which is a little bit faster than in my previous tests).

Oooooh... so that's what would be `UV_THREADPOOL_SIZE`?  
It actually is the size of each chunk sharp treats from the queue?

I've made a test by setting `process.env.UV_THREADPOOL_SIZE = 1`, and it indeed does slows down a lot my script, because each image is treated one after the other. And less cores are used (3 cores against 8 before).

BUT! No matter what value I give to `concurrency`, it doesn't affect my CPU usage. Without changing `UV_THREADPOOL_SIZE`, the CPU is always uses at more than 700% by the node process. So that means sharp is using 8 cores, even when I set `concurrency` to 1.

> The default value of concurrency will already be 8, although this setting will help much less as the bottleneck for processing PNG images will probably be encoding time, which is single-threaded anyway.

When you say that the encoding of PNG images is single-threaded, does it count for all other formats? Because I convert each input image to 9 output images: 3 PNG, 3 WebP and 3 AVIF. From my experience, AVIF is the format that takes the most time to encode.

And if encoding is single-threaded, what is the purpose of `concurrency`? What would be multi-threaded other than image encoding?


Sorry for all those questions but I'm very curious. I'd be glad to help complete the docs with those data.  
And also share some gists in order to help people setting up batch image resizing on their servers.

PS: love your work! When you know the global ecological footprint of loading images on internet, your library is a life saver.

**Comment by nagman** [comment]
Did you have the time to check my last questions?

**Comment by lovell** [comment]
I've added a few more details to the documentation that will hopefully answer your questions - please see commit https://github.com/lovell/sharp/commit/5b03579e5c29dc38e90b0c2bb685411ff89b570e

**Comment by nagman** [comment]
Ooooh, ok I see!
I totally misunderstood what a thread is.
After having watched a bunch of videos that explain the difference between a core and a thread, I got it.

I mean: I _think_ I got it.

Correct me if I'm wrong.

There are two threadpools here:
- the `libuv`'s one: `UV_THREADPOOL_SIZE`
- and the `libvips`' one: `concurrency`

The first one is used for I/O operations, like reading a file (hard drive to memory) or writing a file (memory to hard drive).
The threads are used by libuv to create asynchronous node APIs like `fs`, which are based on synchronous system APIs.
We perform an asynchronous call to `fs.readFile()`, for instance, and it will use one of the threads to delegate the request to the system's filesystem API. The thread will stay in waiting mode, and will end once the system returns the read file.
So if the threadpool size is 8, it can perform 8 read operations in parallel (so 8 image inputs).
As you said: the `UV_THREADPOOL_SIZE` defines how many images are processed in parallel.
But more accurately, it would be "how many images are read or written in parallel".

Then there is the libvips' threadpool size (which is internally used by glib).
It's used for image operations that happen directly in the memory (transform a PNG to an AVIF for instance).
If we set the `concurrency` to 8, it will spawn 8 threads.
Each thread is given a part of the full image.
So we have 8 parts that are treated concurrently (or parallelly?).
Yet, don't know if the threads are spawned on the same core or if they are dispatched between all cores (would be better this way).

In short we have two bottlenecks: I/O operations and image compression.

I'm not sure for `libvips` because I haven't seen any difference between a high `concurrency` and a low one, after having set `effort` to the maximum (so that the bottleneck is not I/O but compression).

That was for the understanding part.

Now in practice, as I will run my script on my personal machine and I don't care if it uses 100% of my cpu power, I can set the `concurrency` to 8 and `UV_THREADPOOL_SIZE` to 8 also.
It will use as much power as it can, and i'm fine with it.
